### ISSUE TITLE
Fail fast, fail smart

### ISSUE THREAD

**Issue by Safihre** [sabnzbd/sabnzbd#1200]
Due to a comment on Reddit I was thinking if we can improve on the failing of jobs that can't be completed.

Since we download the first article of each (non-par2) file at the start of the download to get the filename for Direct Unpack, we could use this info to fail hopless jobs faster. 

So I propose a new mechanism with a switch to enable/disable, enabled by default(?):
_If there's more than 10 non-par2 files and 80% or more of these files the first articles were missing, abort._

Thoughts? @sanderjo @OneCDOnly @jcfp @thezoggy

**Comment by jcfp** [comment]
I've been using this very mechanism to quickly check on an nzb's chances for a while, and in my experience "all first articles green" in the job status always equals an unproblematic download. So yes please!

Obviously, the feature shouldn't be exclusive to jobs that qualify for direct unpack, which is currently the only reason for this behaviour. The proposed >10 non-par2 files with >=80% broken seems safe as well.

**Comment by thezoggy** [comment]
its been forever since I've seen a job fail... so i'm probably not much help on this one.

**Comment by sanderjo** [comment]
> If there's more than 10 non-par2 files and 80% or more of these files the first articles were missing, abort.

Sounds clever. 

Let's check:
- Normal post with a few (1%) missing articles => extremely small chance that that 1% is in the first articles, so this will go allright, and repair will handle it. Good.
- Out of retention => everything is missing => 100% of first articles => fast fail. Good
- Fake post => fast fail. Good.
- DMCA => with deleteof  full post & (almost?) all articles => fast fail. Good. 
- But DMCA with 50% removal of the articles (thus repair not possible), then the 80% broken check will NOT be reached ... not good. 

So ... looks good, except for the last case?

Out of curiosity ... where is the current code that checks the first articles?

**Comment by jcfp** [comment]
> * But DMCA with 50% removal of the articles (thus repair not possible), then the 80% broken check will NOT be reached ... not good. Why the 80%, and not 40%

One cannot know whether that partial DMCA removal includes any percentage of the first articles of the files, could well be that all articles used by this check are present even though there's 50% missing overall. On the other hand, setting the bar lower increases the risk that a very small percentage of random missing articles kills an otherwise perfectly repairable job because those very articles just happen to be the first ones per files.

The current repairability-based check faces similar issues in corner cases. For example, a partial removal with par2 cover about equal to the pct. of articles removed could lead to the last article of a job being decisive (i.e., same result as when not doing any completion checking at all). In reality, it works as intended 99.9% of the time.

**Comment by Safihre** [comment]
With this input I'll give it a go then 👍

**Comment by sanderjo** [comment]
Feedback:

I created a NZB (post size: 666Mb) and malformed all articles (so: non-existing). Details of the NZB:
* 75 files
* 67 rar files
* 8 par2 files 
* Total 1180 articles

SAB (on the special branch) 
* fails after "5% 40 MB Missing articles" (that seems a lot, maybe including the par2 files?). It took some time (5-10 seconds), which seemed long to me. 
* I checked sabnzbd.log, that has 69 "missing from all servers, discarding" lines. Ah, only 2 more than the 67 rar files. That looks good.

![image](https://user-images.githubusercontent.com/1273502/50386998-f859d580-06f2-11e9-9615-870b439648e5.png)
